Rework the definitions of the extern ABIs

ehuss opened the pull request:
This updates the definitions of the extern ABIs to try to make them a little clearer as to what they mean.

This also adds new statements about the equivalence of some ABIs to the "C" ABI.

Fixes https://github.com/rust-lang/reference/issues/1946.

ehuss committed:
Rework the definitions of the extern ABIs

This updates the definitions of the extern ABIs to try to make them a
little clearer as to what they mean.

This also adds new statements about the equivalence of some ABIs
to the "C" ABI.

ehuss commented:
@ChrisDenton We rephrased the description of the `"system"` ABI, and were wondering if you could take a look to see if the new text was accurate and conveyed the right meaning.

@Amanieu or @ChrisDenton  We made a few new claims in this edit about the `"C"` ABI. We were sure who to ask, but were wondering if you knew the answer. In particular it is specifying that:
* "win64" is the same as the "C" ABI on Windows x86_64 targets
* "sysv64" is the same as the "C" ABI on non-Windows x86_64 targets

ChrisDenton commented on src/items/external-blocks.md:
```diff
@@ -111,13 +111,16 @@ r[items.extern.abi.standard]
 
 r[items.extern.abi.system]
-* `unsafe extern "system"` --- Usually the same as `extern "C"`, except on Win32, in which case it's `"stdcall"`, or what you should use to link to the Windows API itself.
+* `unsafe extern "system"` --- This is equivalent to `extern "C"` except on Win32. On Win32, it's equivalent to `"stdcall"`.
```
It's a pre-existing issue but maybe this should spell out "Windows 32-bit" or "Windows x86_32" rather than abbreviating to "Win32"?

RalfJung reviewed (commented):
This sounds right but I don't actually know that much about all these target-specific ABIs.

Not sure how I feel about linking to the Microsoft docs for some of these (even when they are not Microsoft-defined ABIs, or are they?), but I don't know a better place to link to either.

ehuss commented on src/items/external-blocks.md:
```diff
@@ -111,13 +111,16 @@ r[items.extern.abi.standard]
+* `unsafe extern "system"` --- This is equivalent to `extern "C"` except on Windows x86_32 where it is equivalent to `"stdcall"`.
+
+  > [!NOTE]
+  > As the correct underlying ABI on Windows is target-specific, it's best to use `extern "system"` when attemping to link to the Windows API.
```
Updated, though I think you meant cdecl here?

ehuss commented on src/items/external-blocks.md:
```diff
@@ -126,33 +129,57 @@ r[items.extern.abi.platform]
 
 r[items.extern.abi.cdecl]
-* `unsafe extern "cdecl"` --- The default for x86_32 C code.
+* `unsafe extern "cdecl"` --- The calling convention typically used with x86_32 C code.
```
I believe the answer here is also "yes" (assuming the `s` is a typo).

RalfJung commented on src/items/external-blocks.md:
```diff
@@ -126,33 +129,57 @@ r[items.extern.abi.platform]
+
 r[items.extern.abi.stdcall]
-* `unsafe extern "stdcall"` --- The default for the Win32 API on x86_32.
+* `unsafe extern "stdcall"` --- The calling convention typically used by the [Win32 API] on x86_32.
```
Seems worthwhile, doesn't it?

ehuss committed:
Add more ABI correspondence with other compilers

traviscross committed:
Clarify default when not using `extern fn`

Due to changes we've recently merged discussing the default ABI of an
`extern` that doesn't specify one ("C"), the description of "Rust"
being the default ABI for functions that followed immediately after
this read as a bit odd.  Let's revise this to clarify that "Rust" is
the default ABI when `extern fn` is not used.

traviscross committed:
Clarify about Windows functions with explicit ABI

Some Windows API functions have an explicitly defined ABI.  We were
trying to suggest this with wording that was a bit subtle; let's spell
it out explicitly.

traviscross committed:
Convert ABI "details" links to lists

For each of the ABIs, we have outbound links to documents with more
details.  For most of these, we display the URL, since these URLs are
short and readable enough that it's probably more informative to just
show them as compared to trying to describe the document.  However,
the URLs are still long enough that they look better as list items, so
let's switch to doing it that way.

traviscross committed:
Strengthen wording on "C" ABI

When using the "C" ABI, do we "match" the default ABI chosen by the
dominant C compiler for the target or do we "attempt" to match it?

Tough question.  There are believed to be some exceptions in current
behavior.

However, in the Reference, we generally state things in terms of what
the consensus is on how they should be and then note the bugs, so
let's do that here.  We'll just say that we "match" it.
